neo/Ledger/MemoryPool.cs

```diff
@@ -4,162 +4,358 @@
+        /// (_unsortedTransactions, _sortedLowPrioTransactions, and _sortedHighPrioTransactions) after each block.
+        /// </summary>
+        private readonly ConcurrentDictionary<UInt256, PoolItem> _unverifiedTransactions = new ConcurrentDictionary<UInt256, PoolItem>();
+        private readonly SortedSet<PoolItem> _unverifiedSortedHighPriorityTransactions = new SortedSet<PoolItem>();
+        private readonly SortedSet<PoolItem> _unverifiedSortedLowPriorityTransactions = new SortedSet<PoolItem>();
+        
+        private int MaxHighPriorityTxsPerBlock => Settings.Default.MaxTransactionsPerBlock 
+            - Settings.Default.MaxFreeTransactionsPerBlock;
+        
+        /// <summary>
+        /// Total maximum capacity of transactions the pool can hold
+        /// </summary>
         public int Capacity { get; }
-        public int Count => _mem_pool_fee.Count + _mem_pool_free.Count;
+        
+        /// <summary>
+        /// Total count of transactions in the pool
+        /// </summary>
+        public int Count => _unsortedTransactions.Count + _unverifiedTransactions.Count;
 
+        /// <summary>
+        /// Total count of verified transactions in the pool.
+        /// </summary>
+        public int VerifiedCount => _unsortedTransactions.Count;
+        
         public MemoryPool(int capacity)
         {
             Capacity = capacity;
         }
 
-        public void Clear()
+        /// <summary>
+        /// Determine whether the pool is holding this transaction and has at some point verified it.
+        /// Note: The pool may not have verified it since the last block was persisted. To get only the
+        ///       transactions that have been verified during this block use GetVerifiedTransactions()
+        /// </summary>
+        /// <param name="hash">the transaction hash</param>
+        /// <returns>true if the MemoryPool contain the transaction</returns>
+        public bool ContainsKey(UInt256 hash) => _unsortedTransactions.ContainsKey(hash) 
+             || _unverifiedTransactions.ContainsKey(hash);
+        
+        public bool TryGetValue(UInt256 hash, out Transaction tx)
         {
-            _mem_pool_free.Clear();
-            _mem_pool_fee.Clear();
+            bool ret = _unsortedTransactions.TryGetValue(hash, out PoolItem item)
+                       || _unverifiedTransactions.TryGetValue(hash, out item);
+            tx = ret ? item.Transaction : null;
+            return ret;
         }
-
-        public bool ContainsKey(UInt256 hash) => _mem_pool_free.ContainsKey(hash) || _mem_pool_fee.ContainsKey(hash);
-
+        
+        public bool TryGetUnverified(UInt256 hash, out Transaction tx)
+        {
+            bool ret = _unverifiedTransactions.TryGetValue(hash, out PoolItem item);
+            tx = ret ? item.Transaction : null;
+            return ret;
+        }
+        
+        // Note: this isn't used in Fill during consensus, fill uses GetVerifiedTransactions()
```

Comment:
Fill should get the Sorted, right? Function GetVerifiedTransactions is returning unsorted.

Reply:
Yeah, it needs locking to return the transactions sorted. I expected to make that change before this PR is accepted and merged. I'll probably add a R/W lock for it.

Reply:
Made changes. It gets them sorted now for fill.

Reply:
You are right, @erikzhang https://github.com/neo-project/neo-plugins/blob/db04e756a64c149e86d117fe89879c72f319c7df/SimplePolicy/SimplePolicyPlugin.cs#L58

But sort the sorted should be almost O(n), yes? @igormcoelho 
Thus, better to send it sorted anyway.